Mp new ladder line symbols

hbeni opened the pull request:
Implemented fixed-ladder and rope-ladder line types for SKBB (replacing the red line todo reminder currently used).

<s>The symbol works like the ceiling-meander, like discussed on the mailing list.

The symbols work with slopes and straight lines, however optimizations could be still made to make the line more smoother and to save on symbols: currently the outer shafts of the ladder is drawn with small sections. It would be better to draw them using just two single curves following the entire line Path with some offset.
Does someone know how this works? The examples are still somewhat confusing to me...</s>

The smbol works good with straight lines; small linear dents are supported too. With bezier curves there are strange effects, bu most ladders i know are straight anyway ;)
The current code draws the ladder-sides in two paths, which should be more efficient and avoids artefacts with the old code.

hbeni commented:
In commit 44f29d9 i adjusted the code to draw the sides of the ladders in one path. This looks more nicely now and should perform better.

mbudaj commented:
There is a minor issue with ladders drawn as polygonal lines. In the picture you attached the middle segment of the 3-piece ladder is more narrow than the start and end segments. It would be nice if you could fix this before merging.

I also noted that this PR contains some commits which were already merged as part of other PRs.

hbeni commented:
Yes, now that you pointed at it, i can see the difference in the middle segment.
However, sadly i have no idea how i could fix this.
The current algorythm just draws the path shifted -90°/+90° in respect to the start of line, which causes the problem to occur.

As far as one evening googling told me, this is not an easy issue to fix (see also the source code comments).
The obvious solution wold be to retrace the entire path but with an offset. Unfortunately i was not able to achieve this (to few metapost skills yet), but the current solution works good with straight lines (ladders are usually straight; the current workaround if not is to use splitted lines).
I hope this s enough to get the pull included, i hope i can adjust this once i got better in metapost...

As for the other contained commits; i just squashed the relevant commits and rebased the pull request on upstream/master 464b6185d775, hope that helps.

mbudaj commented:
If we assume just linear segments (no curves), the solution should not be that complicated. I would construct path in offset the following way:
- iterate over the points (corners) of the original path
- at each corner determine the direction and distance where to move it
  - bisect the angle A between the segments to find the direction (take it as 180 deg at the beginning and end of the path)
  - find the distance as offset/(sin(A/2))
- just connect the newly created points

It would be better to implement it this way, because the shifting of the whole path would not work if there are e.g. perpendicular twists on it.

hbeni commented:
Thank you for your advice.

Do you have any idea where i may peek, because i have no clue how to implement what you have said... :(

mbudaj commented:
OK, I implemented the fixed ladder (including a generic macro for calculating path in offset from a given polygon). 

I'm not so sure about the rope ladder, especially the "half-circle steps". Is this an Austrian standard?

hbeni commented:
No, no official standard, tough i have seen it this way. I also think that this distinguishes the two ladder types nicely.

Thank you for your work on this, i will inspect your solution to learn from it :)

hbeni commented:
Said this, i think it would also be cool if the rope ladder line would look similar to the point symbol, i.e. with the curved side lines and with straight steps.

mbudaj commented:
Closing this (fixed ladder is implemented; rope ladder would require either reimplementation using polyline_offset macro or a proper implementation visually similar to the point rope-ladder symbol).
